src/SignalR/common/Http.Connections/src/Internal/DefaultBeforeShutdown.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Http.Connections.Features;
+
+namespace Microsoft.AspNetCore.Http.Connections.Internal
+{
+    internal class DefaultBeforeShutdown : IBeforeShutdown
+    {
+        internal List<Func<Task>> Callbacks { get; } = new();
+
+        public IDisposable Register(Func<Task> func)
+        {
+            lock (Callbacks)
+            {
+                Callbacks.Add(func);
+            }
+            return new Disposable(Callbacks, func);
+        }
+
+        public async Task TriggerAsync()
+        {
+            Func<Task>[] callbacks;
+            lock (Callbacks)
+            {
+                callbacks = Callbacks.ToArray();
+                Callbacks.Clear();
+            }
+            foreach (var callback in callbacks)
+            {
+                try
+                {
+                    await callback();
+                }
+                catch { }
```

Comment:
We should report exceptions. I think throwing an AggregateException like CancellationTokenSource.Cancel() and logging at the trigger site makes sense.